enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+- Operator will be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers)
+
+*Note: static resource creation is delegated to CVO (SA, RBAC, etc.)* 
+
+#### CVO managemnet
+
+The operator image will be build and included in every release payload, and expected to be deployed and running at kubernetes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).
+
+Proposed `CCMO` manifests runlevel is `26`.
+
+#### Cloud provider management
+
+Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.
+
+The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, `ServiceMonitors` for metrics for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.
+
+Forked provider repositories will also contain cloud-provider specific code, and a set of binaries for each of them to build and run. Each provider will be build inside a single image, which will be included in the release payload. The responsibility for the operator will be to choose which provider image should be used in a cluster at any moment.
+
+#### CCM migration from KCM pod
+
+Currently the cloud-specific controller loops are running inside the `kcm` pod. We need to preserve leadership over these parts during migration process. This includes migration from the `openshift-kube-controller-manager` namespace to the `openshift-cloud-controller-manager` namespace, which is not possible with current `leader-election` implementation.
+
+Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes the main implementation details for `Lease` resource. This cluster-scoped resource maintains cross-namespaced component leadership, and would help OpenShift to preserve HA during migration from the in-tree to the out-of-tree.
+
+As an alternative to this implementation, the `CCCMO` could be responsible for reading the state of the `kcm` cluster-operator status, and waiting for `Avaliable` status to be `true`, and `Progressing` to become `false`, which would mean that all `nodes` are at a new revision of the `kcm` with the `CCM` controllers disabled (or enabled, depending on the provider and release version). Then the `CCCMO` will start provisioning nessesary `CCM` resources.
```

Comment:
You shouldn't wait for other operator availability. Whoever is first "wins" the lease and is the main process. The way the migration is described it's possible for either of the controller-managers to grab the lock and run.

Reply:
Would kcm give up its lease eventually?

Reply:
This suggestion came from the upstream docs but I don't think it's been worded very clearly here.

The idea is that you: 
- Update KCM to turn off the controllers that will be taken over by CCM.
- Wait for KCM to deploy and become ready, ie, old pods have gone away, new pods are not running the CCM controllers
- Now you can deploy CCM knowing that nothing will conflict

I think the idea was that we could build this logic into the CCCMO so that it could only create the deployment for the CCM once it's sure that KCM is no longer running the cloud controllers. This would involve a small amount of downtime for the cloud controllers though, is that acceptable?

Reply:
It would be good to add just that to this proposal, so anyone looking at it has a clear idea how the migration will look like.